internal/proxy/options.go

```diff
@@ -1,311 +1,197 @@
+		// The key published at this endpoint allows upstreams to decrypt the `Sso-Signature`
+		// header, and validate the integrity and authenticity of a request.
+
+		certs := make(map[string]string)
+		id, key := signer.PublicKey()
+		certs[id] = key
 
-		SkipAuthPreflight: false,
-		RequestLogging:    true,
+		certsAsStr, err := json.MarshalIndent(certs, "", "  ")
+		if err != nil {
+			return fmt.Errorf("could not marshal public certs as JSON: %s", err)
+		}
 
-		DefaultUpstreamTimeout:          time.Duration(1) * time.Second,
-		DefaultUpstreamTCPResetDeadline: time.Duration(1) * time.Minute,
+		op.requestSigner = signer
+		op.publicCertsJSON = certsAsStr
 
-		DefaultAllowedEmailAddresses: []string{},
-		DefaultAllowedEmailDomains:   []string{},
-		DefaultAllowedGroups:         []string{},
-		PassAccessToken:              false,
+		return nil
 	}
 }
 
-// Validate validates options
-func (o *Options) Validate() error {
-	msgs := make([]string, 0)
-	if o.Cluster == "" {
-		msgs = append(msgs, "missing setting: cluster")
-	}
-	if o.ProviderURLString == "" {
-		msgs = append(msgs, "missing setting: provider-url")
-	}
-	if o.UpstreamConfigsFile == "" {
-		msgs = append(msgs, "missing setting: upstream-configs")
-	}
-	if o.CookieSecret == "" {
-		msgs = append(msgs, "missing setting: cookie-secret")
-	}
-	if o.ClientID == "" {
-		msgs = append(msgs, "missing setting: client-id")
-	}
-	if o.ClientSecret == "" {
-		msgs = append(msgs, "missing setting: client-secret")
+// SetUpstreamConfig sets the upstream config as a functional option
+func SetUpstreamConfig(upstreamConfig *UpstreamConfig) func(*OAuthProxy) error {
+	return func(op *OAuthProxy) error {
+		op.upstreamConfig = upstreamConfig
+		return nil
 	}
+}
 
-	if o.StatsdHost == "" {
-		msgs = append(msgs, "missing setting: statsd-host")
+// SetProxyHandler sets the proxy handler as a functional option
+func SetProxyHandler(handler http.Handler) func(*OAuthProxy) error {
+	return func(op *OAuthProxy) error {
+		op.handler = handler
+		return nil
 	}
+}
 
-	if o.StatsdPort == 0 {
-		msgs = append(msgs, "missing setting: statsd-port")
+// SetValidator sets the email validator as a functional option
+func SetValidators(validators []options.Validator) func(*OAuthProxy) error {
+	return func(op *OAuthProxy) error {
+		op.Validators = validators
+		return nil
 	}
-	if o.StatsdHost != "" && o.StatsdPort != 0 {
-		StatsdClient, err := newStatsdClient(o)
-		if err != nil {
-			msgs = append(msgs, fmt.Sprintf("error creating statsd client error=%q", err))
-		}
-		o.StatsdClient = StatsdClient
+}
+
+// SetProvider sets the provider as a functional option
+func SetProvider(provider providers.Provider) func(*OAuthProxy) error {
+	return func(op *OAuthProxy) error {
+		op.provider = provider
+		return nil
 	}
+}
 
-	if o.UpstreamConfigsFile != "" {
-		rawBytes, err := ioutil.ReadFile(o.UpstreamConfigsFile)
+func SetUpstreamConfigs(uc *UpstreamConfigs, cc *CookieConfig, svc ServerConfig) error {
```

Comment:
Looks like `cc` doesn't need to be a pointer as we're only reading a value.

So if you make that change you'll also need to pass `config.SessionConfig.CookieConfig` as a value.

Reply:
Good catch, thank you!